instrumentation-genai/opentelemetry-instrumentation-google-genai/src/opentelemetry/instrumentation/google_genai/dict_util.py

```diff
@@ -0,0 +1,222 @@
+        return _flatten_dict(
+            value,
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if isinstance(value, list):
+        if _is_homogenous_primitive_list(value):
+            return {key: value}
+        return _flatten_list(
+            value,
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if hasattr(value, "model_dump"):
+        return _flatten_dict(
+            value.model_dump(),
+            key_prefix=key,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+    if _from_json:
+        raise ValueError(
+            f"Cannot flatten value with key {key}; value: {value}"
+        )
+    json_value = json.loads(json.dumps(value))
+    return _flatten_value(
+        key,
+        json_value,
+        exclude_keys=exclude_keys,
+        rename_keys=rename_keys,
+        flatten_functions=flatten_functions,
+        _from_json=True,
+    )
+
+
+def _flatten_value(
+    key: str,
+    value: Any,
+    exclude_keys: Set[str],
+    rename_keys: Dict[str, str],
+    flatten_functions: Dict[str, Callable],
+    _from_json=False,
+) -> FlattenedDict:
+    if value is None:
+        return {}
+    key_names = set([key])
+    renamed_key = rename_keys.get(key)
+    if renamed_key is not None:
+        key_names.add(renamed_key)
+        key = renamed_key
+    if key_names & exclude_keys:
+        return {}
+    if _is_primitive(value):
+        return {key: value}
+    return _flatten_compound_value(
+        key=key,
+        value=value,
+        exclude_keys=exclude_keys,
+        rename_keys=rename_keys,
+        flatten_functions=flatten_functions,
+        key_names=key_names,
+        _from_json=_from_json,
+    )
+
+
+def _flatten_dict(
+    d: Dict[str, Any],
+    key_prefix: str,
+    exclude_keys: Set[str],
+    rename_keys: Dict[str, str],
+    flatten_functions: Dict[str, Callable],
+) -> FlattenedDict:
+    result = {}
+    for key, value in d.items():
+        if key in exclude_keys:
+            continue
+        full_key = _concat_key(key_prefix, key)
+        flattened = _flatten_value(
+            full_key,
+            value,
+            exclude_keys=exclude_keys,
+            rename_keys=rename_keys,
+            flatten_functions=flatten_functions,
+        )
+        result.update(flattened)
+    return result
+
+
+def _flatten_list(
+    lst: list[Any],
+    key_prefix: str,
+    exclude_keys: Set[str],
+    rename_keys: Dict[str, str],
+    flatten_functions: Dict[str, Callable],
+) -> FlattenedDict:
+    result = {}
+    result[_concat_key(key_prefix, "length")] = len(lst)
+    for index, value in enumerate(lst):
+        full_key = f"{key_prefix}[{index}]"
```

Comment:
What are the `foo.length` and `foo[i]` for? I don't see those in the test cases anywhere

Reply:
Good point. I've updated the AllowList to be capable of expressing wildcards like `foo.bar[*].baz` and `some.prefix.*` so that this is compatible with the allowlist approach.

Reply:
I think dumping it all to JSON defeats the purpose of flattening and is more likely to run into span attribute size limit issues. Let's leave as-is if you don't mind.

Reply:
Marking resolved.